test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestFeatureService.java

```diff
@@ -10,57 +10,129 @@
-            return historicalFeaturesForVersion == null ? Set.of() : historicalFeaturesForVersion.getValue();
-        }).orElse(allHistoricalFeatures));
-
-        this.historicalFeaturesPredicate = minNodeVersion.<Predicate<String>>map(
-            v -> featureId -> hasHistoricalFeature(historicalFeatures, v, featureId)
-        ).orElse(featureId -> true); // We can safely assume that new non-semantic versions (serverless) support all historical features
+    ESRestTestFeatureService(List<FeatureSpecification> specs, Collection<Version> nodeVersions, Set<String> clusterStateFeatures) {
+        if (MetadataHolder.HISTORICAL_FEATURES != null) {
+            specs.add(MetadataHolder.HISTORICAL_FEATURES);
+        }
+        var historicalFeatures = FeatureData.createFromSpecifications(specs).getHistoricalFeatures();
+        this.knownHistoricalFeatureNames = Optional.ofNullable(historicalFeatures.lastEntry()).map(Map.Entry::getValue).orElse(emptySet());
         this.clusterStateFeatures = clusterStateFeatures;
+        this.allSupportedFeatures = Sets.union(
+            clusterStateFeatures,
+            nodeVersions.stream()
+                .min(Comparator.naturalOrder())
+                .map(v -> Optional.ofNullable(historicalFeatures.floorEntry(v)).map(Map.Entry::getValue).orElse(emptySet()))
+                .orElse(knownHistoricalFeatureNames)
+        );
     }
 
-    private static boolean hasHistoricalFeature(NavigableMap<Version, Set<String>> historicalFeatures, Version version, String featureId) {
-        var features = historicalFeatures.floorEntry(version);
-        return features != null && features.getValue().contains(featureId);
+    public static boolean isLegacyTestFramework() {
+        // Historical features information is unavailable when using legacy test plugins
+        return MetadataHolder.HISTORICAL_FEATURES == null;
     }
 
     @Override
     public boolean clusterHasFeature(String featureId) {
-        if (clusterStateFeatures.contains(featureId)) {
-            return true;
+        if (isLegacyTestFramework() == false
+            && MetadataHolder.FEATURE_NAMES.contains(featureId) == false
+            && knownHistoricalFeatureNames.contains(featureId) == false) {
+            throw new IllegalArgumentException(
+                Strings.format(
+                    "Unknown feature %s: check the feature has been added to the correct FeatureSpecification in the relevant module or, "
+                        + "if this is a legacy feature used only in tests, to a test-only FeatureSpecification such as %s.",
+                    featureId,
+                    RestTestLegacyFeatures.class.getCanonicalName()
+                )
+            );
+        }
+        if (MetadataHolder.FEATURE_NAMES.contains(featureId)) {
+            return clusterStateFeatures.contains(featureId);
         }
-        return historicalFeaturesPredicate.test(featureId);
+        return allSupportedFeatures.contains(featureId);
```

Comment:
nit: it is a bit wasteful to check first clusterStateFeatures then (again) in all supported features. Also, would this work wit the legacy test framework?
Maybe skip line 81-83 and just tests on allSupportedFeatures?

Reply:
Happy to just test against `allSupportedFeatures`. 

However, the current code is functionally different and ignores historical features that are eventually known as real features. Not sure if that's validated somewhere else. In any case, that's why I kept that code as is.

Reply:
Yes, there is some kind of "precedence", so if a feature is defined with the same name, "real" features takes precedence. But does it really change?
Anyway, happy to maintain the current behaviour (but you will need to remove the `if (MetadataHolder.FEATURE_NAMES.contains(featureId))`, that does not work in all cases)

Reply:
Right, thanks! That might also trigger the ESQL issues. How to I actually recognise the "legacy test framework"?

Reply:
Actually, getting confused myself, with the legacy framework the check on FEATURE_NAMES always fails, so it's always checking against `allSupportedFeatures`. This should work as expected, right?

Reply:
Yes, you are right. It's kind of useless, but since it is an emptyset when you do not have feature metadata, it should do no harm. I would simplify the code nonetheless (so you don't have to do all the guesswork we did :) ) but it's a minor/optional change.

Reply:
I've removed the check against `clusterStateFeatures` in favor of always checking `allSupportedFeatures` 👍